src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs

```diff
@@ -654,7 +654,7 @@ internal unsafe bool WriteEvent(ref EventDescriptor eventDescriptor, IntPtr even
         }
 
         /// <summary>Workaround for inability to stackalloc object[EtwAPIMaxRefObjCount == 8].</summary>
-        private struct EightObjects
+        private ref struct EightObjects
         {
             internal object? _arg0;
```

Comment:
This should be InlineArray. Can InlineArrays be `ref structs`?

Reply:
It would result in a warning:
```
warning CS9184: 'Inline arrays' language feature is not supported for an inline array type that is not valid as a type argument, or has element type that is not valid as a type argument.
```

Reply:
@AaronRobinsonMSFT is this supported in the runtime?

Reply:
> This should be InlineArray. Can InlineArrays be `ref structs`?

Yes.

> @AaronRobinsonMSFT is this supported in the runtime?

As in, can `ref struct`s be marked as `InlineArray` or something else? If the former, see above.

Reply:
Applied `InlineArray` with `#pragma warning disable CS9184`

Reply:
> As in, can ref structs be marked as InlineArray or something else? 

I mean in the runtime. Basically want to make sure it's fully supported there before we look into potentially changing the language too.

Reply:
Yes, applying `InlineArray` to a ByRefLike type is a supported scenario from the runtime perspective.

Reply:
>IMO this diagnostic message is out-of-date as in C# 13 we have added the support for byref-like type arguments.

Language support for inline arrays depends on several generic APIs exposed by BCL. The inline array type must be a valid type argument for some of them. If those APIs were all updated with `allows ref struct` constraint, we probably could support a `ref struct` inline array which doesn't use a `ref`/`ref struct` as its element. However, the scenario feels pathological to me. I.e. the inline array is marked as a `ref struct` not for the purpose of storing references in it. Therefore, I am not sure if it is worth the effort to enable this partial support. We are still unable to support a `ref struct` as an element of an inline array because span types do not allow a `ref struct` as a type argument.

Reply:
> Language support for inline arrays depends on several generic APIs exposed by BCL. The inline array type must be a valid type argument for some of them. If those APIs were all updated with allows ref struct constraint

Which APIs in particular?

Reply:
> Which APIs in particular?

Unless I missing something, for this specific scenario the list is limited to ```Unsafe.As<,>``` and ```Unsafe.AsRef<>```. As I mentioned above, for complete support of ref structs, we need span types that allow a ref struct as a type argument.

Reply:
> for this specific scenario the list is limited to Unsafe.As<,> and Unsafe.AsRef<>

Ok, those are already annotated accordingly:
https://github.com/dotnet/runtime/blob/20f11b0505d018c345446d06e6aeaeb64befb1e5/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs#L87-L89
https://github.com/dotnet/runtime/blob/20f11b0505d018c345446d06e6aeaeb64befb1e5/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs#L696-L697

Reply:
Is this one of the cases though where we have explicit guidance that `Unsafe.As` is safe to use in the compiler? Basically is @jkotas signing off that use extending `Unsafe.As` where the source is a `ref struct` in the `InlineArray` case is supported?

Reply:
> extending Unsafe.As where the source is a ref struct in the InlineArray case is supported

Yes, that's fine. It does not fundamentally change the special sequence to create Span from InlineArray that has to work forever.

(I would have preferred if that special sequence did not use Unsafe.As - https://github.com/dotnet/csharplang/pull/7064#discussion_r1167048749 , but that ship sailed.)

Reply:
Is that a place where we could define a new runtime helper method and adjust compiler to use that going forward? Won't fix the existing debt but going forward if there was a better helper to use we could move to it.

Reply:
Something like:
```csharp
public class RuntimeHelpers
{
    public static Span<Element> InlineArrayAsSpan<InlineArray, Element>(ref InlineArray inlineArray) where InlineArray: allows ref struct, Element: allows ref struct;
}
```
? I agree that it looks better than the current solution with `Unsafe.As`. It is definitely possible; it is a bunch of work to teach all codegens to expand it as intrinsic or throw exception if there is any sort of mismatch.

Reply:
Opened https://github.com/dotnet/runtime/issues/105586